### ISSUE TITLE
ODF: unknown conductivity unit ratio; should be 'ratio', 'mS/cm' or 'S/m'

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#1185]
See file here (renamed to txt extension):
[CTD_DGR2009001_1_1_DN.txt](https://github.com/dankelley/oce/files/742285/CTD_DGR2009001_1_1_DN.txt)

It gives the warnings:
```r
> d <- read.oce('CTD_DGR2009001_1_1_DN.ODF')
Warning messages:
1: In ODFNames2oceNames(ODFnames, ODFunits, PARAMETER_HEADER = NULL,  :
  unable to interpret ODFunits[2]='metres', for item named 'depth'
2: In ODFNames2oceNames(ODFnames, ODFunits, PARAMETER_HEADER = NULL,  :
  unable to interpret ODFunits[4]='ratio', for item named 'conductivity'
3: In read.odf(file = file, columns = columns, debug = debug - 1) :
  "CRAT_01" should be a conductivity ratio, but setting unit to "ratio" since that is in the data file; see ?read.odf for an example of rectifying this unit error.
> plotProfile(d, xtype='conductivity')
Error in plotProfile(d, xtype = "conductivity") : 
  unknown conductivity unit ratio; should be 'ratio', 'mS/cm' or 'S/m'
> str(d@metadata$units)
List of 6
 $ pressure    :List of 2
  ..$ unit :  expression(dbar)
  ..$ scale: chr ""
 $ depth       :List of 2
  ..$ unit :  expression("metres")
  ..$ scale: chr "metres"
 $ temperature :List of 2
  ..$ unit :  expression(degree * C)
  ..$ scale: chr "ITS-90"
 $ conductivity:List of 2
  ..$ unit :  expression("ratio")
  ..$ scale: chr "ratio"
 $ salinity    :List of 2
  ..$ unit :  expression()
  ..$ scale: chr "PSS-78"
 $ sigmat      :List of 2
  ..$ unit :  expression(kg/m^3)
  ..$ scale: chr ""
```

But the units **are** `ratio`?

Any idea what's going on @dankelley ?

**Comment by dankelley** [comment]
I'll look this aft, thanks.

**Comment by richardsc** [comment]
I *think* this has to do with the following lines:
```r
        } else {
            warning("unable to interpret ODFunits[", i, "]='", ODFunits[i], "', for item named '", names[i], "'", sep="")
            list(unit=as.expression(ODFunits[i]), scale=ODFunits[i])
        }
```
where there is a difference between: `expression(ratio)` and `expression("ratio")` (note the quotes). Do you know how to create the former from a character string stored in an object?

The easiest fix is to add a grep for "ratio" to the big list of unit greps and set it manually, but this is inelegant in my opinion (as the above code chunk *should* have taken care of this).

**Comment by richardsc** [comment]
Actually, I'm not sure that the above comment about `expression(ratio)` vs`expression("ratio")` matters, since I've tried changing the coercion to:
```r
            list(unit=parse(text=ODFunits[i]), scale=ODFunits[i])
```
based on this [SO post](http://stackoverflow.com/questions/1743698/evaluate-expression-given-as-a-string). But trying to do `plotProfile(d, xtype='conductivity')` still kicks up the same error:
```
> d <- read.odf('CTD_DGR2009001_1_1_DN.ODF')
Warning message:
In read.odf("CTD_DGR2009001_1_1_DN.ODF") :
  "CRAT_01" should be a conductivity ratio, but setting unit to "ratio" since that is in the data file; see ?read.odf for an example of rectifying this unit error.
> plotProfile(d, xtype='conductivity')
Error in plotProfile(d, xtype = "conductivity") : 
  unknown conductivity unit ratio; should be 'ratio', 'mS/cm' or 'S/m'
```

So, the problem is maybe in `plotProfile()`. Looking there now.

**Comment by dankelley** [comment]
As a clue, I see
```R
> plot(d)
> plot(as.ctd(d))
> plotProfile(d, xtype='conductivity')
Error in plotProfile(d, xtype = "conductivity") : 
  unknown conductivity unit ratio; should be 'ratio', 'mS/cm' or 'S/m'
Called from: plotProfile(d, xtype = "conductivity")
Browse[1]> 
```
so if we make it a `ctd` first, then things are ok.

I can take over on this, Clark ... I see I'm assigned and I'm at a computer now.

**Comment by dankelley** [comment]
Also, I see
```R
> summary(d)
ODF Summary
-----------

* Instrument:          SBE
* Instrument model:    SBE 25
* Instr. serial no.:   26
* File source:         /Users/kelley/Dropbox/CTD_DGR2009001_1_1_DN.ODF
* Institute:           BIO/Coastal Ocean Sciences
* Chief scientist:     Jim Hamilton
* Date:                2009-08-07 12:20:00
* Start time:          2009-08-07 12:20:00
* Cruise:              Barrow Strait - Arctic Ocean Through Flow 2009
* Vessel:              CCGS Des Groseilliers
* Station:             1
* Statistics of data

                            Min.    Mean    Max.    Dim. OriginalName
pressure [dbar]             2.5     24      45.5    87   PRES_01     
depth [m]                   2.474   23.75   45.024  87   DEPH_01     
temperature [°C, ITS-90]    -1.4622 -0.8174 0.1664  87   TEMP_01     
conductivity [ratio, ratio] 0.59581 0.59786 0.60347 87   CRAT_01     
salinity [PSS-78]           30.71   31.403  31.96   87   PSAL_01     
sigmat [kg/m³]              24.639  25.233  25.702  87   SIGT_01     

* Processing Log
    - 2017-01-31 14:25:47 UTC: `create 'odf' object`
    - 2017-01-31 14:25:48 UTC: `read.odf(file = "~/Dropbox/CTD_DGR2009001_1_1_DN.ODF")`
```

**Comment by richardsc** [comment]
Hang tight. I re-assigned to me and have been making some changes.

**Comment by richardsc** [comment]
Unfortunately I wasn't paying attention and did them on `develop` instead of a separate branch. I'm just going to push and then you can look at the commits and revert if you want.

**Comment by richardsc** [comment]
Fixed (I think) in commit c864182b57ed48108132a20a263358ca65094081.

Note that there were some man page updates after I built the package, so those got pushed separately in commit 1f17229fd1a1d862c70e6222f4f9cf7ace4f54b0

**Comment by dankelley** [comment]
Did it complain when you pushed? I see as below. I can fix up, but you could too..?
```
git pull
Updating c864182b..1f17229f
error: Your local changes to the following files would be overwritten by merge:
	man/as.ctd.Rd
	man/ctd-class.Rd
	man/ctdTrim.Rd
Please commit your changes or stash them before you merge.
Aborting
```

**Comment by richardsc** [comment]
No, no complaint. But I don't know why those changes occurred ... try pulling again?

**Comment by dankelley** [comment]
I'm just going to stash my changes (and ignore the stash) because it seems a few extra whitespaces at the ends of lines.  I'm rebuilding locally, and I'll remove this and the past few comments when that's ok, because it's just a distraction to anyone coming into this thread to learn about the bug.

**Comment by dankelley** [comment]
This seems to work great for me. However, I'm worried about how `plotProfile(as.ctd())` worked, whereas `plotProfile()` failed. I propose we leave this open until that's sorted out ... Clark, you can reassign me to do that, if you like.  (You could also just close this ... I plan to check into this confusion tonight, anyway.)

**Comment by richardsc** [comment]
Reassigned to you. I think it would be good to see why the `as.ctd()` made a different, but you know that code better than I.

**Comment by dankelley** [comment]
Thanks. My worry is that `as.ctd(d)` might be *ignoring* the unit in `d`, and that we only found this out with this case of a brokenly-framed unit.

**Comment by dankelley** [comment]
I've had a look at this idea that results differ if you first do `as.ctd()` (see https://github.com/dankelley/oce/issues/1185#issuecomment-276448711 above) and ... well ... I do not think this is actually the case.  I made a [test issue](https://github.com/dankelley/oce-issues/tree/master/11xx/1185) for this, and tried with (a) the existing code and (b) the code if I remove the change Clark made earlier today in `R/odf.R` line 808 (in which a test is made as to whether `ustring != "ratio"`). Lo and behold, I get identical results. I think I was just plain wrong when I wrote https://github.com/dankelley/oce/issues/1185#issuecomment-276448711 and so I am now going to close this. 

Summary: closing because `develop` commit 4076fdc2a68ad7f4ee6a779bc8c4e3b3ba8a5c70 seems fine.
